### ISSUE TITLE
Exception while executing wordcount example

### ISSUE THREAD

**Issue by rmetzger** [stratosphere/stratosphere#857]
Hi,
while testing the current `master` branch to see if we can release it this way, I found an issue with the new Java API / Wordcount example.

```
05/24/2014 21:56:32:    Job execution switched to status FAILED
Error: The program execution failed: java.lang.ArrayIndexOutOfBoundsException: -128
    at java.util.ArrayList.elementData(ArrayList.java:400)
    at java.util.ArrayList.get(ArrayList.java:413)
    at eu.stratosphere.nephele.io.RuntimeOutputGate.writeRecord(RuntimeOutputGate.java:268)
    at eu.stratosphere.nephele.io.AbstractRecordWriter.emit(AbstractRecordWriter.java:92)
    at eu.stratosphere.pact.runtime.shipping.OutputCollector.collect(OutputCollector.java:82)
    at eu.stratosphere.example.java.wordcount.WordCount$Tokenizer.flatMap(WordCount.java:101)
    at eu.stratosphere.example.java.wordcount.WordCount$Tokenizer.flatMap(WordCount.java:91)
    at eu.stratosphere.pact.runtime.task.chaining.ChainedFlatMapDriver.collect(ChainedFlatMapDriver.java:71)
    at eu.stratosphere.pact.runtime.task.DataSourceTask.invoke(DataSourceTask.java:219)
    at eu.stratosphere.nephele.execution.RuntimeEnvironment.run(RuntimeEnvironment.java:351)
    at java.lang.Thread.run(Thread.java:745)

For a more detailed error message use the vebose output option '-v'.
```

I successfully executed the Scala and Record Wordcount implementations with the same configuration (cluster size, data set).
The error is reproducible.

**Comment by StephanEwen** [comment]
Bug in the hash partitioner.

**Comment by rmetzger** [comment]
The bug is caused by a the following problem: 
`Integer.MIN_VALUE == -Integer.MIN_VALUE`. Also, `Math.abs(Integer.MIN_VALUE) = Integer.MIN_VALUE`.
See also here: http://stackoverflow.com/questions/18565485/why-is-absolute-of-integer-min-value-equivalent-to-integer-min-value

The error only occurs if the hash value is accidentally `Integer.MIN_VALUE`.

The RecordOutputEmitter is doing some hash-salting magic:

``` java
for (int i = 0; i < DEFAULT_SALT.length; i++) {
  hash ^= ((hash << 5) + DEFAULT_SALT[i] + (hash >> 2));
}
```

I don't know if this is preventing this case. 
How can we fix this?

**Comment by uce** [comment]
Good catch!

The easiest thing will be to test for MIN_VALUE?

**Comment by rmetzger** [comment]
Yeah, this will cause a slight skew in the data. Also I'm not sure about the performance impact. This is on the "critical path".

**Comment by uce** [comment]
PS: This will have an impact on performance, but is there any other way to do it with less skew?

**Comment by rmetzger** [comment]
I tested two variants to solve this issue

``` java
    public static void main(String[] args) {
        int numberOfChannels = 208;
        int a;

        long start = System.currentTimeMillis();
        int c;
        for(int i = Integer.MIN_VALUE; i < Integer.MAX_VALUE; i++) {
            c = ((i % numberOfChannels) + numberOfChannels)  % numberOfChannels;
            if(c < 0 ) {
                System.err.println("wrong");
            }
        }
        System.err.println("First test "+ (System.currentTimeMillis() - start) );

        long start2 = System.currentTimeMillis();
        for(int i = Integer.MIN_VALUE; i < Integer.MAX_VALUE; i++) {
            if(i < 0) {
                if(i == Integer.MIN_VALUE) {
                    a = Integer.MAX_VALUE % numberOfChannels;
                } else {
                    a = -i % numberOfChannels;
                }
            } else {
                a= i % numberOfChannels;
            }
            if(a < 0 ) {
                System.err.println("wrong");
            }
        }

        System.err.println("Second test "+ (System.currentTimeMillis() - start2) );
    }
```

So for `((i % numberOfChannels) + numberOfChannels)  % numberOfChannels;`
the runtime is 33980

for the `if if else else` variant, I have 18505.
I ran the tests with Java 8.

**Comment by StephanEwen** [comment]
Something that branches, but follows the same branch almost always should
be faster than modulo, if I recall correctly. Divisions are expensive,
unless they are dividing by powers of two.

**Comment by uce** [comment]
I agree. So let's special case the MIN_VALUE and increment the hash in that case.

@rmetzger Do you want to do it or should I? Let's make sure to include a small test.

New RC after the PR?

**Comment by rmetzger** [comment]
I would like to see how https://github.com/stratosphere/stratosphere/issues/860 develops. I will not create a new RC until all known issues are addressed.

**Comment by uce** [comment]
At the time the issue was not known ;-)

**Comment by fhueske** [comment]
Fixed with 8677ad9ad03cfaa37e654ebd8bd868928fbddff2 and c9589b82d671e0afe76affd01c2e2c71b151a667
